Step2단계 미션 풀리퀘입니다~!

hjlee83 opened the pull request:
Step2단계 미션 풀리퀘입니다~!

hojinDev committed:
docs : readme.md

기능 요구사항과 관련된 내용을 README.md에 작성.

hojinDev committed:
feat(lotto): 로또 구매 기능 추가

로또샵에서 금액을 입력받아 로또를 구매할 수 있는 기능을 추가. 1000원 단위가 아닌 금액 입력 시 예외 처리 추가.

test: 로또 구매 기능에 대한 테스트 코드 작성.

yebink-woowahan committed:
feat : 로또 금액을 담는 Money 클래스 생성

hojinDev committed:
feat(lotto): 로또 생성 및 구매 후 번호 출력 기능 추가

로또 생성 클래스와 구매 후 번호를 출력하는 기능을 추가. 로또 구매 시 로또 객체를 반환하도록 변경하고, 테스트 코드 수정 및 추가.

footer: JVM Toolchain 17로 버전 변경.

yebink-woowahan committed:
feat : 지난주 당첨 금액 입력 코드 작성

hojinDev committed:
feat(lotto): 당첨 결과 통계 및 수익률 계산 기능 추가

당첨 번호와 구매한 로또를 비교해 당첨 결과 통계 및 수익률을 계산하는 기능 추가.
Rank enum 생성으로 당첨 등수 및 상금을 관리하고 LottoResult 클래스에서 결과를 처리하도록 구현.

footer: 테스트 코드 및 README.md 업데이트.

yebink-woowahan committed:
refactor : 테스트 리팩토링 (BehaviorSpec -> FreeSpec)

yebink-woowahan committed:
refactor : 함수 라인 10줄 이내로 맞추기, 매직넘버 상수화

hojinDev committed:
refactor(lotto): 상수 위치를 companion object로 이동

로또 클래스의 가독성과 책임 분리를 위해 상수들을 companion object로 이동.
이를 통해 상수가 클래스 인스턴스와 무관한 공용 정보임을 명확히 표현.

footer: 코드 정리 및 클래스 구조 개선.

hojinDev committed:
fix(lotto): 예외 메시지에 상수 사용하도록 수정

예외 메시지에 상수를 활용하여 코드 중복을 제거하고 유지 보수성을 개선.
상수 값이 변경되더라도 메시지가 자동으로 반영되도록 처리.

footer: 예외 처리 가독성 향상.

hojinDev committed:
refactor(lotto): 금액 검증 로직 개선

금액 검증 로직을 init 및 invoke로 분리하여 가독성과 재사용성을 향상.
에러 메시지 및 최소 금액 상수를 추가하여 코드 중복 제거 및 유지 보수성 개선.

footer: 로직 재구성으로 유효성 검사 효율성 증가.

hojinDev committed:
refactor(lotto): 금액 파싱 및 검증 로직 개선

문자열 금액 입력 시 초기화 로직을 개선하고 중복된 검증 코드를 parseAndValidate 함수로 통합.
init 블록과 생성자에서 최소 금액 상수를 활용하여 코드 간소화.

footer: 코드 재사용성 및 유지 보수성 향상.

hojinDev committed:
refactor(lotto): buyLotto 메서드명을 sellLotto로 변경

메서드의 동작과 의도를 더 명확히 표현하기 위해 buyLotto를 sellLotto로 리팩터링.
이에 따라 관련 호출부와 테스트 코드도 동일하게 수정.

footer: 가독성 및 코드 일관성 강화.

hojinDev committed:
docs(readme): 로또 보너스 볼 및 수동 번호 입력 요구사항 추가

로또 보너스 볼 입력 조건과 수동 번호 입력 기능 관련 요구사항을 README.md에 추가.
프로그램 구현 및 유지 보수를 위한 상세 가이드라인 포함.

footer: README.md 업데이트로 요구사항 명확화.

hojinDev committed:
docs(readme): 1차 피드백 항목 추가

피드백 내용을 반영하여 출력 기능 분리를 위한 View 클래스 제안 항목을 README.md에 추가.
요구사항 명확화를 통해 구현 가이드라인 강화.

footer: README.md 업데이트로 요구사항 보완.

hojinDev committed:
refactor(lotto): LottoTicket 클래스 도입 및 관련 코드 리팩터링

로또 구매 데이터를 캡슐화하기 위해 LottoTicket 클래스를 도입.
LottoShop 및 LottoResult 클래스에서 LottoTicket을 활용하도록 수정.
로또 개수와 금액 관련 로직을 LottoTicket으로 이동하여 코드의 책임 분리 및 가독성 향상.

footer: 관련 테스트 코드 수정 및 예외 메시지 통일.

hojinDev committed:
refactor(lotto): LottoNumber 클래스 도입 및 관련 코드 리팩터링

로또 번호를 캡슐화하기 위해 LottoNumber 클래스를 도입하고 Lotto 및 WinLotto에서 이를 활용하도록 수정.
로또 번호 검증 로직을 LottoNumber로 이동하여 책임 분리 및 중복 코드 제거.

footer: LottoTest와 LottoResultTest 코드 수정 및 예외 메시지 통일.

hojinDev committed:
refactor(lotto): 번호 범위 상수를 LottoNumber로 이동

로또 번호 범위를 나타내는 상수를 LottoNumber로 이동하여 책임을 분리.
Lotto 클래스와 LottoNumber 간의 참조를 정리해 코드 가독성 및 유지 보수성을 향상.

footer: 관련 테스트 코드 및 예외 메시지 수정 완료.

hojinDev committed:
feat(lotto): 보너스 볼 로직 추가 및 관련 코드 리팩터링

로또 당첨 결과에 보너스 볼 로직을 반영하여 2등 판별 기능을 추가.
Rank 및 LottoResult 클래스에서 보너스 매칭 로직 구현. Lotto 및 LottoTicket 생성 방식을 개선.

footer: 관련 테스트 및 README.md 요구사항 반영 완료.

hojinDev committed:
feat(lotto): 로또 번호 유효성 검증 로직 추가

로또 번호가 6개의 서로 다른 숫자로 구성되어야 한다는 유효성 검증 로직을 init 블록에 추가.
유효하지 않은 경우 IllegalArgumentException을 발생시키도록 구현.

footer: 관련 테스트 케이스 추가 및 예외 메시지 검증 완료.

hojinDev committed:
feat(lotto): 수동 로또 구매 기능 추가

수동으로 구매할 로또 수와 번호를 입력받아 로또를 생성하는 기능을 LottoTicket에 추가.
이에 따라 관련 클래스 및 메서드를 리팩터링하고, LottoShop 판매 로직을 개선.

footer: 관련 테스트 및 README 업데이트 완료.

hojinDev committed:
feat(lotto): 당첨 번호 및 보너스 볼 입력 로직 개선

당첨 번호와 보너스 볼 입력 방식을 WinLotto 및 BonusBallLotto의 companion object로 이동시켜 리팩터링.
입력 유효성 검증과 예외 처리 로직을 통합하여 코드 중복을 제거하고 가독성을 향상.

footer: 관련 테스트 코드 수정 및 LottoResult 출력 로직 업데이트 완료.

hjlee83 committed:
Merge branch 'hjlee83' into step2

men7627 commented on src/main/kotlin/lotto/domain/BonusBallLotto.kt:
```diff
@@ -0,0 +1,29 @@
+        require(bonusNumber !in winLotto.winLotto.numbers) {
+            "보너스 번호는 당첨 번호와 겹칠 수 없습니다."
+        }
+    }
```
꼼꼼한 유효성 체크.. 굿입니다!
(저는 깜빡했는데.. 적용해야 겠군요)

men7627 commented on src/main/kotlin/lotto/BonusBallLotto.kt:
```diff
@@ -0,0 +1,29 @@
+                    println(e.message)
+                }
+            }
+        }
```
지금 수정하기엔 수정 범위가 너무 커서 그냥 의견만 드려보자면,,
I/O 관련한 기능들은 별도로 구분하는 것이 좀 더 객체지향적이고, 도메인로직과 I/O 로직이 분리되어 깔끔해지지 않을까 싶습니다.

men7627 commented on src/main/kotlin/lotto/Lotto.kt:
```diff
@@ -1,21 +1,28 @@
-class Lotto(input: List<Int>) {
-    val numbers = input
-
+data class Lotto(val numbers: List<LottoNumber>) {
```
요러한 일급컬렉션 같은 형태의 객체 data class를 만들어도 될지 의문입니다.
data class는 setter 가 생성되는데, 사실상 세터에도 유효성 검증이 필요할 것 같아서 고렇게 생각했습니다.

setter 를 override 할 수도 있는 것으로 아는데 그 방법도 고려해보시면 좋을 것 같습니다.
하지만 저라면 data class를 사용하지 않을 것 같습니다!

hjlee83 replied:
value 클래스로 해야했는데 어제 만들때만 해도 제 무지로 비롯한 이슈네요 ㅎㅎㅎ변경하겠습니다.

men7627 commented on src/main/kotlin/lotto/domain/Lotto.kt:
```diff
@@ -1,21 +1,28 @@
-            if (!(MINIMUM_LOTTO_NUMBER..MAXIMUM_LOTTO_NUMBER).contains(it)) {
-                throw IllegalArgumentException("${MINIMUM_LOTTO_NUMBER}부터 ${MAXIMUM_LOTTO_NUMBER}까지의 숫자를 입력하세요")
-            }
+        require(numbers.toSet().size == LOTTO_NUMBER_COUNT) {
+            "다른 숫자 6개가 아니면 안됩니다."
```
역할을 분리해서 유효성 검증도 최소단위로 분리가 된 것 같아서 좀 더 객체지향스러워진 것 같습니다! 👍

men7627 commented on src/main/kotlin/lotto/LottoGame.kt:
```diff
@@ -1,38 +1,23 @@
-        buyLotto()
-    }
-    return listOf()
+fun buyLotto(): LottoTicket {
+    val money = Money.read()
+    val manualCount = ManualCount.read(money)
+    val manualLottoOrder = ManualLottoOrder.create(manualCount)
+    val lottoTicket = LottoShop.sellLotto(money, manualLottoOrder)
+    lottoTicket.print()
+    return lottoTicket
```
I/O 를 각각의 객체에 포함시켰군요!
좀 더 객체지향스럽고 깔끔해진 것 같습니다!

men7627 commented on src/main/kotlin/lotto/LottoNumber.kt:
```diff
@@ -0,0 +1,16 @@
+package lotto
+
+data class LottoNumber(val number: Int) {
```
마찬가지로 data class 사용하는 것이 맞을까? 고민해보면 좋을 것 같습니다!

hjlee83 replied:
요것도 value 클래스로 해야했는데 변경하겠습니다.

men7627 commented on src/main/kotlin/lotto/domain/LottoNumber.kt:
```diff
@@ -0,0 +1,16 @@
+        }
+    }
+
+    constructor(number: String) : this(number.toInt())
```
```
companion object {
    fun fromString(number: String) = LottoNumber(number.toInt())

    const val MINIMUM = 1
    const val MAXIMUM = 45
}
```
와 같이 factory 메소드로 만들면 메서드에 의미를 부여할 수도 있을 것 같습니다! (단순 의견)

men7627 commented on src/main/kotlin/lotto/input/ManualLottoOrderInputView.kt:
```diff
@@ -0,0 +1,61 @@
+        private const val ERROR_MESSAGE_INVALID = "올바른 로또 번호를 입력하세요"
+
+        fun create(manualCount: ManualCount): ManualLottoOrder {
+            while (true) {
+                try {
+                    val manualLottoNumbers =
+                        if (manualCount.count > 0) {
+                            println("수동으로 구매할 로또 번호를 입력해 주세요.")
+                            (1..manualCount.count).joinToString("\n") { readln() }
+                        } else {
+                            null
+                        }
+                    return ManualLottoOrder(manualCount, manualLottoNumbers)
+                } catch (e: IllegalArgumentException) {
+                    println(e.message)
+                }
```
while > try > if 로 하나의 메서드 depth가 깊어지는 것 같은데 이를 해소하면 가독성이 좀 더 좋아질 것 같습니다.

men7627 commented on src/test/kotlin/lotto/MoneyTest.kt:
```diff
@@ -39,9 +39,8 @@ class MoneyTest : FreeSpec({
-                shouldThrow<IllegalArgumentException> { Money(input) }
-            exception.message shouldBe "올바른 금액을 입력하세요"
+            val money = Money(input)
+            money.price shouldBe 0
```
요 유효성은 왜 제거하셨나요~?

hjlee83 replied:
수동으로 돌렸을때의 금액을 요걸로 해서 0원이라는 money 가 허용되어야했습니다. ZeroMoney 라는 클래스를 만들어도 되었겠지만 요것까지 하기에는 좀 과하다 느껴졌어요.

men7627 commented on src/test/kotlin/lotto/domain/RankTest.kt:
```diff
@@ -13,4 +13,12 @@ class RankTest : FreeSpec({
         val rank = Rank.valueOf(2)
         rank shouldBe Rank.MISS
     }
+    "5개가 일치하고 보너스 번호가 있으면 2등이 나온다." {
+        val rank = Rank.valueOf(5, hasBonus = true)
+        rank shouldBe Rank.SECOND
+    }
+    "5개가 일치하고 보너스 번호가 없으면 3등이 나온다." {
+        val rank = Rank.valueOf(5, hasBonus = false)
+        rank shouldBe Rank.THIRD
+    }
```
요런 포맷으로도 테스트를 짤 수 있군요!
깔끔하고 좋은 것 같습니다. 💯

men7627 commented on src/test/kotlin/lotto/domain/LottoTest.kt:
```diff
@@ -1,13 +1,41 @@
+        exception.message shouldBe "다른 숫자 6개가 아니면 안됩니다."
+    }
+
+    "다른 숫자 6개가 아니면 안된다." {
+        val exception =
+            shouldThrow<IllegalArgumentException> {
+                Lotto(
+                    listOf(
+                        LottoNumber(1),
+                        LottoNumber(1),
+                        LottoNumber(1),
+                        LottoNumber(1),
+                        LottoNumber(1),
+                        LottoNumber(1),
+                    ),
+                )
+            }
```
1개만 중복, 2개만 중복, ... , 케이스가 다양할 것 같은데
forAll 과 row 를 사용해서 다양한 파라미터를 한번에 검증해보는 테스트를 짜보셔도 좋을 것 같습니다.

```
    "덧셈 테스트" {
        forAll(
            row(1, 2, 3),
            row(10, 5, 15),
            row(-1, 5, 4)
        ) { a, b, expected ->
            (a + b) shouldBe expected
        }
    }
```

men7627 commented on src/test/kotlin/lotto/domain/LottoResultTest.kt:
```diff
@@ -6,10 +6,27 @@ import io.kotest.matchers.shouldBe
 class LottoResultTest : FreeSpec({
     "로또 당첨 결과 및 수익률 테스트" - {
```
Given, When, Then 구문을 추가하면 좀 더 가독성이 좋을 것 같습니다!

men7627 commented on src/main/kotlin/lotto/domain/LottoShop.kt:
```diff
@@ -1,19 +1,17 @@
+        ): LottoTicket {
+            require(money.price % LOTTO_UNIT_PRICE == 0) { ERROR_MESSAGE_INVALID_UNIT }
+            val autoLottoCount = (money.price - manualLottoOrder.usedMoney.price) / LOTTO_UNIT_PRICE
+            return LottoTicket(money, manualLottoOrder.lottos, List(autoLottoCount) { Lotto.createRandom() })
```
아주 코틀린스러운 문법이군요!

men7627 commented on src/main/kotlin/lotto/domain/LottoResult.kt:
```diff
@@ -12,26 +16,39 @@ class LottoResult(val winLotto: WinLotto, val lottos: List<Lotto>) {
+            val matchCount = winLotto.matchCount(lotto)
+            val hasBonus = bonusBallLotto.matches(lotto)
+            val rank = Rank.valueOf(matchCount, hasBonus)
             matchMap[rank] = matchMap.getOrDefault(rank, 0) + 1
```
좀 더 깔끔하게 변환 가능할 것 같습니다!
요런식으로
```
matchMap.merge(rank, 1, Int::plus)
```

men7627 commented on README.md:
```diff
@@ -8,6 +8,26 @@
 - [x] 지난 주 당첨 번호를 입력할 수 있어야 한다.
     - [x] 1 ~ 45 사이의 중복 없는 6개 숫자가 아닐 경우, 안내 문구와 함께 재시도할 수 있도록 한다.
     - [x] 숫자는 스페이스바로 구분한다.
+- [x] 보너스 볼을 입력해야한다.
+    - [x] 지난 주 담청 번호와 중복되면 안된다.
+    - [x] 1 ~ 45 숫자여야한다.
 - [x] 당첨 통계 결과를 출력한다.
     - [x] 당첨 통계 갯수와 금액 정보는 enum을 활용해서 생성한다.
     - [x] 당첨 통계 판독기에서 로또 번호와 당첨 번호를 비교해 당첨 통계 / 수익률을 생성한다.
+- [x] 수동으로 추첨 번호를 입력할 수 있도록 해야 한다.
+    - [x] 수동으로 구매할 로또 수를 입력해야한다.
+    - [x] 수동으로 구매한 로또 수만큼 구매할 번호를 입력해야한다.
+    - [x] 숫자는 스페이스바로 구분한다.
+    - [x] 금액에서 수동으로 구매한 로또 수를 제외한 만큼은 자동으로 구매해야한다.
```
👍

men7627 reviewed (changes requested):
고생하셨습니다 호진님 ~~ 💯 💯
1단계에서 드렸던 피드백까지 반영해주셨군요! 👍
코드가 좀 더 객체지향스러워진 것 같습니다,
간단한 의견들 드렸는데, 참고해주시면 좋을 것 같습니다!

hojinDev committed:
refactor(lotto): Lotto 및 LottoNumber를 value class로 변경

Lotto와 LottoNumber를 value class로 전환하여 메모리 사용 효율성을 개선하고 최적화.
기존 데이터 검증 로직은 유지하면서 구조를 간소화.

footer: 관련 테스트 코드 및 초기화 로직 검증 완료.

hojinDev committed:
feat(lotto): 금액 입력 검증 및 처리 기능 분리

금액 입력 및 유효성 검증 로직을 MoneyInputView로 분리하여 책임을 명확화.
관련 테스트 코드 MoneyInputViewTest 추가, Money 클래스에서는 최소 값 검증만 수행하도록 간소화.

footer: 테스트 코드 추가

hojinDev committed:
feat(lotto): 수동 구매 입력 처리 로직 분리

ManualCount의 수동 구매 입력 처리 코드를 ManualCountInputView로 분리하여 책임 명확화.
관련하여 유효성 검증 로직과 예외 처리 로직 통합. 관련 테스트 코드 추가.

footer: ManualCountInputViewTest 추가 및 기존 코드 리팩터링 완료.

hojinDev committed:
refactor(lotto): 도메인 및 입력 관련 패키지 구조 개편

lotto의 파일들을 lotto.domain 및 lotto.input 패키지로 분리하여 책임 명확화.
Lotto, Money 등 주요 클래스를 domain 패키지로 이동하고, 입력 뷰 관련 코드를 input 패키지로 이동.
관련 테스트 코드의 패키

hojinDev committed:
feat(lotto): 수동 로또 번호 입력 뷰 및 검증 로직 추가

수동 로또 번호 입력 처리를 담당하는 ManualLottoOrderInputView 클래스를 추가.
기존 ManualLottoOrder의 입력 및 검증 로직을 해당 클래스로 분리하여 책임 명확화.
관련 테스트 코드 추가로 입력 유효성 검증 로직 검증 완료.

footer: ManualLottoOrderInputViewTest, ManualLottoOrderTest 추가 및 기존 코드 리팩터링.

hojinDev committed:
feat(lotto): 당첨 번호 입력 및 검증 로직 리팩터링

WinLotto 입력 로직을 WinLottoInputView로 분리하고 책임 명확화.
관련하여 입력 예외 처리 로직 통합 및 LottoMessage

hojinDev committed:
refactor(lotto): WinLotto 클래스 제거 및 관련 코드 정리

WinLotto 클래스의 불필요한 역할 제거를 위해 해당 클래스를 삭제.
matchCount 로직은 Lotto 도메인에서 처리하도록 구조 단순화 및 코드 중복 방지.

footer: 불필요한 클래스 제거를 통해 가독성과 유지 보

hojinDev committed:
feat(lotto): 보너스 볼 입력 및 로또 결과 출력 로직 개선

보너스 볼 입력과 검증을 처리하는 BonusBallInputView 추가.
LottoResult 출력 전용 LottoResultOutputView로 결과 출력 책임 분리.
이로 인해 LottoResult 초기화 로직 단순화 및 메인 함수 코드 가독성 개선.

footer: 테스트 코드 추가 및 기존 로직 리팩터링 완료.

men7627 reviewed (approved):
리뷰를 좀 많이 드린 것 같은데, 잘 반영해주셔서 감사합니다!
코드가 좀 더 객체지향스러워진 것 같습니다,
data class 사용 안하는게 좋겠다고만 했지, value class 는 몰랐는데 덕분에 하나 또 알아갑니다!
고생하셨습니다! 💯
